assets/javascripts/lib/discourse-markdown/discourse-chat-transcript.js

```diff
@@ -0,0 +1,159 @@
+let customMarkdownCookFn;
+
+const chatTranscriptRule = {
+  tag: "chat",
+
+  replace: function (state, tagInfo, content) {
+    const options = state.md.options.discourse;
+    let [username, messageIdStart, messageTimeStart] =
+      (tagInfo.attrs.quote && tagInfo.attrs.quote.split(";")) || [];
+    let channelName = tagInfo.attrs.channel;
+
+    if (!username || !messageIdStart || !messageTimeStart) {
+      return;
+    }
+
+    let wrapperDivToken = state.push("div_chat_transcript_wrap", "div", 1);
+    wrapperDivToken.attrs = [["class", "discourse-chat-transcript"]];
+    wrapperDivToken.attrs.push(["data-message-id", messageIdStart]);
+    wrapperDivToken.attrs.push(["data-username", username]);
+    wrapperDivToken.attrs.push(["data-datetime", messageTimeStart]);
+
+    if (channelName) {
+      wrapperDivToken.attrs.push(["data-channel-name", channelName]);
+      let metaDivToken = state.push("div_chat_transcript_meta", "div", 1);
+      metaDivToken.attrs = [["class", "chat-transcript-meta"]];
+      const channelToken = state.push("html_inline", "", 0);
+      channelToken.content = I18n.t("chat.quote.original_channel", {
+        channel: `#${channelName}`,
+      });
+      state.push("div_chat_transcript_meta", "div", -1);
+    }
+
+    let userDivToken = state.push("div_chat_transcript_user", "div", 1);
+    userDivToken.attrs = [["class", "chat-transcript-user"]];
+
+    // start: user avatar
+    let avatarDivToken = state.push(
+      "div_chat_transcript_user_avatar",
+      "div",
+      1
+    );
+    avatarDivToken.attrs = [["class", "chat-transcript-user-avatar"]];
+
+    // server-side, we need to lookup the avatar from the username
+    let avatarImg;
+    if (options.lookupAvatar) {
+      avatarImg = options.lookupAvatar(username);
+    }
+    if (avatarImg) {
+      const avatarImgToken = state.push("html_inline", "", 0);
+      avatarImgToken.content = avatarImg;
+    }
+
+    state.push("div_chat_transcript_user_avatar", "div", -1);
+    // end: user avatar
+
+    // start: username
+    let usernameDivToken = state.push("div_chat_transcript_username", "div", 1);
+    usernameDivToken.attrs = [["class", "chat-transcript-username"]];
+
+    let displayName;
+    if (options.formatUsername) {
+      displayName = options.formatUsername(username);
+    } else {
+      displayName = username;
+    }
+
+    const usernameToken = state.push("html_inline", "", 0);
+    usernameToken.content = displayName;
+
+    state.push("div_chat_transcript_username", "div", -1);
+    // end: username
+
+    // start: time + link to message
+    let datetimeDivToken = state.push("div_chat_transcript_datetime", "div", 1);
+    datetimeDivToken.attrs = [["class", "chat-transcript-datetime"]];
+
+    let linkToken = state.push("link_open", "a", 1);
+    linkToken.attrs = [
+      ["href", options.getURL(`/chat/message/${messageIdStart}`)],
+    ];
+    linkToken.block = false;
+
+    let datetimeToken = state.push("html_inline", "", 0);
```

Comment:
Update: I realised in my convalescence that I can't do this, it will bake the date + time format server side which is not what we want. Need a cooked post decorator client-side for this instead.